warehouse/slave.go

```diff
@@ -189,126 +207,170 @@ func (job *Payload) getDiscardsTable() string {
 }
 
-func (jobRun *JobRun) getLoadFilePath(tableName string) string {
-	job := jobRun.job
+func (jr *JobRun) getLoadFilePath(tableName string) string {
+	job := jr.job
 	randomness := misc.FastUUID().String()
-	return strings.TrimSuffix(jobRun.stagingFilePath, "json.gz") + tableName + fmt.Sprintf(`.%s`, randomness) + fmt.Sprintf(`.%s`, warehouseutils.GetLoadFileFormat(job.LoadFileType))
+	return strings.TrimSuffix(jr.stagingFilePath, "json.gz") + tableName + fmt.Sprintf(`.%s`, randomness) + fmt.Sprintf(`.%s`, warehouseutils.GetLoadFileFormat(job.LoadFileType))
 }
 
 func (job *Payload) getColumnName(columnName string) string {
 	return warehouseutils.ToProviderCase(job.DestinationType, columnName)
 }
 
-type loadFileUploadJob struct {
-	tableName  string
-	outputFile encoding.LoadFileWriter
-}
-
-type loadFileUploadOutput struct {
-	TableName             string
-	Location              string
-	TotalRows             int
-	ContentLength         int64
-	StagingFileID         int64
-	DestinationRevisionID string
-	UseRudderStorage      bool
-}
+// uploadLoadFiles returns the upload output for each file uploaded to object storage
+func (jr *JobRun) uploadLoadFiles(ctx context.Context) ([]loadFileUploadOutput, error) {
+	ctx, cancel := context.WithTimeout(ctx, jr.slaveUploadTimeout)
+	defer cancel()
 
-func (jobRun *JobRun) uploadLoadFilesToObjectStorage(ctx context.Context) ([]loadFileUploadOutput, error) {
-	job := jobRun.job
-	uploader, err := job.getFileManager(job.DestinationConfig, job.UseRudderStorage)
+	uploader, err := jr.job.getFileManager(jr.job.DestinationConfig, jr.job.UseRudderStorage)
 	if err != nil {
-		return []loadFileUploadOutput{}, err
-	}
-	// var loadFileIDs []int64
-	var loadFileUploadOutputs []loadFileUploadOutput
-
-	// take the first staging file id in upload
-	// TODO: support multiple staging files in one upload
-	stagingFileId := jobRun.job.StagingFileID
-
-	loadFileOutputChan := make(chan loadFileUploadOutput, len(jobRun.outputFileWritersMap))
-	loadFileUploadTimer := jobRun.timerStat("load_file_upload_time")
-	uploadJobChan := make(chan *loadFileUploadJob, len(jobRun.outputFileWritersMap))
-	// close chan to avoid memory leak ranging over it
-	defer close(uploadJobChan)
-	uploadErrorChan := make(chan error, numLoadFileUploadWorkers)
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	for i := 0; i < numLoadFileUploadWorkers; i++ {
-		go func(ctx context.Context) {
-			for uploadJob := range uploadJobChan {
+		return nil, fmt.Errorf("creating uploader: %w", err)
+	}
+
+	type result struct {
+		uploadOutput loadFileUploadOutput
+		err          error
+	}
+
+	type uploadJob struct {
+		tableName  string
+		outputFile encoding.LoadFileWriter
+	}
+
+	var output []loadFileUploadOutput
+	var totalUploadTime time.Duration
+	var mu sync.Mutex
+
+	defer func() {
+		if totalUploadTime > 0 {
+			jr.timerStat("load_file_upload_time").SendTiming(totalUploadTime)
+		}
+	}()
```

Comment:
1 staging file is broken down into multiple load files, I thought the error `Metrics with the same name and label values` was caused because we are sending the metrics multiple times from multiple goroutines. So, instead of sending metrics for each file, we can just send the total upload time metrics?

Reply:
```GO
	defer func() {
		if err == nil {
			jr.timerStat("load_file_total_upload_time").SendTiming(totalUploadTime.Load())
		}
	}()
```

Reply:
Yes I understand why you redesigned this but I don't think that is the culprit. Also, we should be able to send multiple timings for the same timer concurrently. The underlying framework should handle that.

From a metrics perspective it is correct to measure how long it takes to upload each file. I think we should keep that like it was before so that we can have more useful metrics in our dashboards. Having the total is a plus and we could consider it if we think it is useful and we cannot derive it with a promql query.